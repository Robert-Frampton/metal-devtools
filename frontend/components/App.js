import Component, {Config} from 'metal-jsx';
import {bindAll, keys} from 'lodash';

import * as constants from '../../shared/constants';
import InitialWarning from './InitialWarning';
import ResizeDivider from './ResizeDivider';
import StatePane from './StatePane';
import TreeNode from './TreeNode';

class App extends Component {
	created() {
		bindAll(this, 'handleFreezeToggle', 'handleResize', 'processMessage');

		this.props.port.onMessage.addListener(this.processMessage);

		this._pendingFlashRemovals = [];
	}

	addFlash(node) {
		if (node && node.classList) {
			node.classList.add('flash');
		}
	}

	addRootComponent(root) {
		this.state.rootComponents = {
			...this.state.rootComponents,
			[root.id]: root
		};
	}

	checkIfRootDetached(id) {
		const roots = this.state.rootComponents;

		if (roots[id]) {
			delete roots[id];
		}

		this.state.rootComponents = roots;
	}

	flashNode(id) {
		const node = document.querySelector(`[data-nodeid="${id}"]`);

		if (node) {
			this.addFlash(node);

			if (!this.state.freezeUpdates) {
				setTimeout(() => this.removeFlash(node), 100);
			} else {
				this._pendingFlashRemovals.push(node);
			}
		}
	}

	handleFreezeToggle({target}) {
		const {checked} = target;

		this.state.freezeUpdates = checked;

		if (!checked) {
			this._pendingFlashRemovals.forEach(node => this.removeFlash(node));

			this._pendingFlashRemovals = [];
		}
	}

	handleResize({clientX}) {
		this.state.firstColumnWidth = clientX;
	}

	processMessage({data, type}) {
		switch (type) {
			case constants.DETACHED:
				this.checkIfRootDetached(data.id);
				break;
			case constants.NEW_ROOT:
				this.addRootComponent(data);
				break;
			case constants.RENDERED:
				this.flashNode(data);
				break;
			case constants.SELECTED:
				this.state.selectedComponent = data;
				break;
			case constants.UPDATE:
				this.updateRootComponent(data);
				break;
			default:
				console.log(`Unknown Message Type: ${type}`);
		}
	}

	removeFlash(node) {
		if (node && node.classList) {
			node.classList.remove('flash');
		}
	}

	resetRoots() {
		this.state.rootComponents = {};
	}

	updateRootComponent(root) {
		const roots = this.state.rootComponents;

		if (roots[root.id]) {
			roots[root.id] = root;
		}

		this.state.rootComponents = roots;
	}

	render() {
		const {
			props: {highlightDOM, inspectDOM, onSelectedChange, setStateFn},
			state: {
				firstColumnWidth,
				freezeUpdates,
				rootComponents,
				selectedComponent
			}
		} = this;

		const rootComponentKeys = keys(rootComponents);

		return (
			<div class="app-container">
				{!rootComponentKeys ||
					(rootComponentKeys &&
						!rootComponentKeys.length &&
						<InitialWarning />)}

				{rootComponentKeys &&
					!!rootComponentKeys.length &&
					<div
						class="roots-wrapper"
						style={firstColumnWidth && `flex-basis:${firstColumnWidth}px;`}
					>
						<div
							class="options"
							title="Freezes highlights by expanded components"
							style={`width: ${firstColumnWidth + 1}px;`}
						>
							<label for="freezeUpdates">{'Freeze Updates'}</label>

							<input
								checked={freezeUpdates}
								name="freezeUpdates"
								onChange={this.handleFreezeToggle}
								type="checkbox"
							/>
						</div>

						{rootComponentKeys.map((key, i) => (
							<TreeNode
								componentNode={rootComponents[key]}
								depth={0}
								key={i}
								highlightDOM={highlightDOM}
								onInspectDOM={inspectDOM}
								onNodeSelect={onSelectedChange}
								selectedId={selectedComponent.id}
							/>
						))}
					</div>}

				<ResizeDivider onResize={this.handleResize} />

				<StatePane
					component={selectedComponent}
					onInspectDOM={inspectDOM}
					setStateFn={setStateFn}
				/>
			</div>
		);
	}
}

App.PROPS = {
	highlightDOM: Config.func(),
	inspectDOM: Config.func(),
	onSelectedChange: Config.func(),
	port: Config.any(),
	setStateFn: Config.func()
};

App.STATE = {
	firstColumnWidth: Config.number(),
	freezeUpdates: Config.value(false),
	rootComponents: Config.value({}),
	selectedComponent: Config.value({})
};

export default App;
