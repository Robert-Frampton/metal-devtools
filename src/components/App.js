import Component, {Config} from 'metal-jsx';
import {Iterable, fromJS} from 'immutable';
import {isBoolean, isNull, isObject} from 'lodash';

import TreeNode from './TreeNode';

const port = chrome.extension.connect({name: 'metal-devtools'});

class App extends Component {
	created() {
		this.selectedChange = this.selectedChange.bind(this);
	}

	attached() {
		port.onMessage.addListener(
			(component) => {
				const {id, remove} = component;

				let retVal = this.state.rootComponents;

				if (remove) {
					delete retVal[id];
				}
				else if (id) {
					retVal = {
						...retVal,
						[id]: component
					};
				}

				this.state.rootComponents = retVal;
			}
		);
	}

	processValue(value) {
		let retVal = value;

		if (isObject(value)) {
			if (Iterable.isIterable(value)) {
				value = fromJS(value);
			}

			retVal = [<pre>{JSON.stringify(value, null, 2)}</pre>];
		}
		else if (isBoolean(value)) {
			retVal = value.toString();
		}
		else if (value === '') {
			retVal = '\'\'';
		}
		else if (isNull(value)) {
			retVal = 'null';
		}

		return retVal;
	}

	selectedChange(node) {
		this.state.selectedComponent = node;
	}

	render() {
		const {rootComponents, selectedComponent} = this.state;

		const rootComponentKeys = Object.keys(rootComponents);

		const dataKeys = Object.keys(selectedComponent.data || {});

		return (
			<div class="container">
				{rootComponentKeys && !rootComponentKeys.length &&
					<div class="warn">
						<div>
							{'If you do not see your components here, try refreshing the page while keeping this panel open.'}
							<br />
							<i>{'Or there may be no Metal.js components on the page.'}</i>
						</div>
					</div>
				}

				{rootComponentKeys && !!rootComponentKeys.length &&
					<div class="nodes">
						{
							rootComponentKeys.map(
								(key, i) => (
									<TreeNode
										componentNode={rootComponents[key]}
										key={i}
										selectedComponent={selectedComponent}
										onNodeClick={this.selectedChange}
									/>
								)
							)
						}
					</div>
				}

				<div class="config">
					<h1>Component Data:</h1>

					{!!dataKeys.length &&
						dataKeys.map(
							dataKey => {
								const state = JSON.parse(selectedComponent.data[dataKey]);
								const stateKeys = Object.keys(state || {});

								return (
									<div>
										<h2>{`${dataKey}:`}</h2>

										<ul>
											{
												stateKeys.map(
													key => (
														<li>
															<b>{`${key}: `}</b>
															<span>{this.processValue(state[key])}</span>
														</li>
													)
												)
											}
										</ul>
									</div>
								);
							}
						)
					}

					{!dataKeys.length &&
						<div>
							<i>{'No Component Data'}</i>
						</div>
					}
				</div>
			</div>
		);
	}
}

App.STATE = {
	rootComponents: Config.value({}),
	selectedComponent: Config.value({})
};

export default App;
