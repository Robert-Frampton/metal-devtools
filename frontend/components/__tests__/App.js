jest.disableAutomock();

import App from '../App';

describe('App', () => {
	it('should render', () => {
		const component = new App({
			port: {
				onMessage: {
					addListener: jest.fn()
				}
			}
		});

		expect(snap(component)).toMatchSnapshot();
	});
});
