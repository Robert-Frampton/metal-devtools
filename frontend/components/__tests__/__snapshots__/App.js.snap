exports[`App should detach 1`] = `
<div class="app-container light-theme">
  <div class="initial-warning-container" key="_metal_jsx_1">
    <div key="_metal_jsx_1">If you do not see your components here, try refreshing the page while keeping this panel open.
      <br key="_metal_jsx_1">
      <i key="_metal_jsx_2">Or there may be no Metal.js components on the page. metal@^2.6.2 required.
        <a href="https://github.com/bryceosterhaus/metal-devtools" target="blank_" key="_metal_jsx_1">Repo</a>
      </i>
    </div>
  </div>
  <div class="resize-container" key="_metal_jsx_3"></div>
  <div class="state-pane-container" key="_metal_jsx_4">
    <div class="header" key="_metal_jsx_1">Component: </div>
    <div key="_metal_jsx_3">
      <i key="_metal_jsx_1">No Component Data</i>
    </div>
  </div>
</div>
`;

exports[`App should render 1`] = `
<div class="app-container light-theme">
  <div class="roots-wrapper" key="_metal_jsx_2">
    <div style="" class="options" title="Freezes highlights by expanded components" key="_metal_jsx_1">
      <label for="freezeUpdates" key="_metal_jsx_1">Freeze Updates</label>
      <input name="freezeUpdates" type="checkbox" key="_metal_jsx_2">
    </div>
    <div class="tree-container" key="_metal_jsx_2">
      <div class="node-wrapper   " style="padding-left: 12px;" key="_metal_jsx_1">
        <span data-nodeid="foo" class="node-name-container" key="_metal_jsx_2">&lt;
          <span class="name" key="_metal_jsx_1"></span> /&gt;</span>
      </div>
    </div>
  </div>
  <div class="resize-container" key="_metal_jsx_3"></div>
  <div class="state-pane-container" key="_metal_jsx_4">
    <div class="header" key="_metal_jsx_1">Component: </div>
    <div key="_metal_jsx_3">
      <i key="_metal_jsx_1">No Component Data</i>
    </div>
  </div>
</div>
`;

exports[`flattenIds should flatten object into an array 1`] = `
Array [
  1,
  2,
  5,
  4,
]
`;
